proposals/mpi-operator-proposal.md

```diff
@@ -0,0 +1,224 @@
+## Motivation
+Kubeflow currently supports distributed training of TensorFlow models using
+[tf-operator](https://github.com/kubeflow/tf-operator), which relies on
+centralized parameter servers for coordination between workers. An alternative
+is a decentralized approach in which workers communicate with each other
+directly via the MPI `allreduce` primitive, without using parameter servers.
+NVIDIA's [NCCL](https://developer.nvidia.com/nccl) library has implemented most
+of the MPI primitives efficiently on GPUs, and Uber's
+[Horovod](https://github.com/uber/horovod) makes it simple to do multi-gpu and
+multi-node training using TensorFlow; the TensorFlow team is also working on an
+`allreduce`
+[implementation](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/contrib/distribute).
+Compared to parameter servers, this approach arguably is more
+[bandwidth optimal](http://www.cs.fsu.edu/~xyuan/paper/09jpdc.pdf) and
+[scales better](https://eng.uber.com/horovod/).
+
+However, right now there is no easy way to launch this type of training jobs on
+Kubernetes. By providing a CRD and a custom controller, we can make
+allreduce-style distributed training as simple as training on a single node.
+
+## Goals
+* Provide a common Custom Resource Definition (CRD) for defining a single-gpu,
+multi-gpu, or multi-node training job.
```

Comment:
I'm guessing "multi-gpu" here means multiple gpus on on node. I don't really know best practices in kubernetes (I'm new here!), but why not use a single container for this with all the GPU devices you want to use whitelisted?

Reply:
That's exactly what we are doing. When you have 8-GPU nodes and run an 8-GPU job, we launch a single pod and rely on mpi/horovod/nccl to parallelize the training across the 8 GPUs (internally `mpirun` forks a new process for each GPU). The basic idea is a single interface can be used to train on as many GPUs as you have available, regardless of node boundaries.

Reply:
This is reasonable, it's a good design choice, makes things simpler.